homeassistant/components/crownstone/hub.py

```diff
@@ -0,0 +1,157 @@
+            websession=aiohttp_client.async_get_clientsession(self.hass),
+        )
+        # Login
+        try:
+            await self.cloud.async_initialize()
+        except CrownstoneAuthenticationError as auth_err:
+            _LOGGER.error(
+                "Auth error during login with type: %s and message: %s",
+                auth_err.type,
+                auth_err.message,
+            )
+            return False
+        except CrownstoneUnknownError as unknown_err:
+            _LOGGER.error("Unknown error during login")
+            raise ConfigEntryNotReady from unknown_err
+
+        # set the sphere we chose to setup in the flow
+        self.sphere = self.cloud.spheres.find(self.config_entry.data[CONF_SPHERE])
+
+        # Create uart manager to manage usb connections
+        # uart.is_ready() returns whether the usb is ready or not.
+        self.uart_manager = UartManager()
+        self.uart_manager.start()
+
+        # Create SSE instance
+        self.sse = CrownstoneSSE(customer_email, customer_password)
+        self.sse.set_access_token(self.cloud.get_access_token())
+        self.sse.start()
+
+        # subscribe to Crownstone ability updates
+        self.sse.add_event_listener(EVENT_ABILITY_CHANGE_DIMMING, self.update_ability)
+        self.sse.add_event_listener(
+            EVENT_ABILITY_CHANGE_SWITCHCRAFT, self.update_ability
+        )
+        self.sse.add_event_listener(
+            EVENT_ABILITY_CHANGE_TAP_TO_TOGGLE, self.update_ability
+        )
+
+        # create listener for when home assistant is stopped
+        self.hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, self.async_stop)
+
+        # register crownstone entities
+        self.hass.async_create_task(
+            self.hass.config_entries.async_forward_entry_setup(
+                self.config_entry, LIGHT_PLATFORM
+            )
+        )
+
+        return True
+
+    async def async_reset(self) -> bool:
+        """
+        Reset the hub after entry removal.
+
+        Config flow will ensure the right email and password are provided.
+        If an authentication error still occurs, return.
+
+        If the setup was successful, unload forwarded entry.
+        """
+        # reset RequestHandler instance
+        self.cloud.reset()
+        # stop uart
+        self.uart_manager.stop()
+        # stop sse client
+        await self.sse.async_stop()
+
+        # authentication failed
+        if self.cloud.spheres is None:
+            return True
+
+        # unload all platform entities
+        result = await self.hass.config_entries.async_forward_entry_unload(
+            self.config_entry, LIGHT_PLATFORM
+        )
+
+        return result is not False
+
+    @callback
+    def update_ability(self, ability_event: AbilityChangeEvent) -> None:
+        """Update the ability information."""
+        # make sure the sphere matches current.
+        update_sphere = self.cloud.spheres.find_by_id(ability_event.sphere_id)
```

Comment:
Is this doing I/O ?

Reply:
No I/O, it looks for a specific ID in dictionary data